### PR TITLE
Preserve empty scoped quota array through load, merge, and sidebar normalization

### PR DESCRIPTION
**Pull request by iceteaSA**
Rebased onto current `main` (the scoped weekly quota limits feature is now merged). This reduces the PR to two correctness fixes on top of that feature — both are latent bugs in the current implementation.

## 1. Empty `scoped: []` is dropped on load and lost in the runtime-state merge

`fetchOAuthQuotaSnapshot` always returns `scoped` as an array, but two paths drop it when empty, which matters because an empty array is a meaningful state (the API returned no model-scoped carve-out) distinct from "no scoped data":

- **`normalizeQuota`** did `if (scoped.length) quota.scoped = scoped`. When a persisted quota has an empty `scoped` and no `five_hour`/`seven_day`, the whole quota object then collapses to `undefined` via the `Object.keys(quota).length` guard. Fixed with an unconditional `quota.scoped = scoped` inside the existing `Array.isArray(value.scoped)` guard (so pre-feature data with no `scoped` key is unaffected).
- **`mergeAccountRuntimeState`** compares freshness via `quotaSnapshotCheckedAt`, which returned `0` for a windowless `{scoped:[]}` snapshot — so the merge treated it as stale and resurrected the previous (e.g. exhausted) scoped window instead of writing the empty array. Fixed by adding a top-level `checkedAt` to `OAuthQuotaSnapshot` (stamped in `fetchOAuthQuotaSnapshot`, preserved in `normalizeQuota`, included in `quotaSnapshotCheckedAt`'s `max`). For a normal snapshot the top-level stamp equals the per-window stamps, so the freshness comparison is unchanged; a windowless empty-scoped snapshot now carries its real timestamp.

Routing is unaffected: an empty `scoped: []` behaves identically to `undefined` in `quotaSnapshotModelScopeIsExhausted` / `getScopedQuotaWindowForModel` (`.find()` over `[]` and the optional-chain over `undefined` both yield no window), and no routing predicate reads the new top-level `checkedAt`. A regression test locks this invariant.

## 2. Collapsed sidebar shows `5h: — 7d: —` placeholders when only a scoped window is visible

`getCollapsedQuotaSummary` unconditionally emitted the `5h`/`7d` segment even when both are null, producing e.g. `5h: — 7d: — Fa: 100%`. Now the `5h`/`7d` segment is emitted only when at least one is present; when both are null (reachable only because a scoped window is visible) the line is just the scoped segment. The all-windows-present and single-present (dash-for-absent) cases are byte-identical.

## Tests

Red-first on the current code for every fix: empty-`[]` round-trip (scoped-only and alongside `five_hour`), the windowless-`{scoped:[]}` merge-freshness repro, top-level `checkedAt` round-trip, the sidebar `normalizeAccountQuota` empty-array case, the collapsed-summary placeholder omission, plus a routing-invariant lock (`scoped: []` is not treated as exhausted). All suites green — core 6, opencode 768, pi 47, e2e 2; typecheck and lint clean.

<details><summary><h3>Greptile Summary</h3></summary>

This PR surfaces model-scoped weekly quota carve-outs (e.g., the Fable-only promo window) by parsing `limits[]` from `/api/oauth/usage`, persisting them as `quota.scoped[]`, and rendering them in the collapsed summary alongside the existing 5h/7d windows. Scoped windows are intentionally excluded from routing, policy, and killswitch logic.

- **Parse & persist**: `mapScopedWeeklyLimits` filters for `kind:\"weekly_scoped\"` + `group:\"weekly\"`, requires a non-empty `display_name`, deduplicates by model id/slug, and always returns an array (never `undefined`). A new top-level `checkedAt` on `OAuthQuotaSnapshot` ensures windowless empty-scoped snapshots win freshness comparisons in `mergeAccountRuntimeState` instead of being treated as stale (T=0).
- **Render**: `getCollapsedQuotaSummary` now omits the `\"5h: — 7d: —\"` prefix when both windows are absent and only scoped data is visible; the fix from the previous thread is present and tested.
- **Isolation**: `quotaSnapshotPassesPolicy` and `killswitchPassesPolicy` iterate only `['five_hour', 'seven_day']`; a fully-exhausted scoped window does not mark an account dead.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge. The additive nature of the change (scoped windows are parsed and displayed but never gate routing or killswitch decisions) limits blast radius, and the freshness/merge logic for empty-scoped snapshots is correctly guarded by the new top-level checkedAt field.

All changed paths are additive or narrowing: the existing five_hour/seven_day parse/persist/render paths are untouched, policy and killswitch checks are explicitly constrained to those two windows, and the new top-level checkedAt correctly handles the windowless-empty-scoped merge scenario. Edge cases (malformed limits, non-finite percents, absent display_name, empty arrays through save/load) are all handled and tested. No routing, auth, or data-loss risk is introduced.

No files require special attention. The tui.tsx rendering changes noted in previous review comments are outside this diff; the padding alignment concern there is tracked separately.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/core/src/accounts.ts | Adds OAuthQuotaSnapshot.checkedAt, propagates scoped: [] through normalizeQuota, and adds top-level checkedAt to fetchOAuthQuotaSnapshot; freshness logic is correct for windowless snapshots. |
| packages/opencode/src/sidebar-state.ts | Fixes collapsed summary to omit '5h: — 7d: —' prefix when only scoped windows are present; normalizeAccountQuota now preserves scoped: [] as intended. |
| packages/opencode/src/tests/accounts.test.ts | Adds round-trip, merge-freshness, policy-exclusion, and empty-array preservation tests; cases are well-chosen and cover the introduced edge conditions. |
| packages/opencode/src/tests/sidebar-state.test.ts | Adds tests for scoped-only collapsed summary, partial-dash primary segment with scoped, and empty-array normalization; all assert the right behavior. |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["/api/oauth/usage response"] --> B["fetchOAuthQuotaSnapshot"]
    B --> C["mapUsageWindow x2"]
    B --> D["mapScopedWeeklyLimits"]
    D --> E["AccountScopedQuotaWindow[]"]
    E --> F["scoped: []"]
    B --> G["checkedAt: now"]
    C --> H["OAuthQuotaSnapshot"]
    F --> H
    G --> H
    H --> I["normalizeQuota"]
    I --> J["OAuthAccount.quota"]
    J --> K["quotaSnapshotCheckedAt"]
    K --> L["mergeAccountRuntimeState"]
    J --> M["projectQuotaToSidebar"]
    M --> N["getCollapsedQuotaSummary"]
    J --> R["quotaSnapshotPassesPolicy / killswitchPassesPolicy"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["/api/oauth/usage response"] --> B["fetchOAuthQuotaSnapshot"]
    B --> C["mapUsageWindow x2"]
    B --> D["mapScopedWeeklyLimits"]
    D --> E["AccountScopedQuotaWindow[]"]
    E --> F["scoped: []"]
    B --> G["checkedAt: now"]
    C --> H["OAuthQuotaSnapshot"]
    F --> H
    G --> H
    H --> I["normalizeQuota"]
    I --> J["OAuthAccount.quota"]
    J --> K["quotaSnapshotCheckedAt"]
    K --> L["mergeAccountRuntimeState"]
    J --> M["projectQuotaToSidebar"]
    M --> N["getCollapsedQuotaSummary"]
    J --> R["quotaSnapshotPassesPolicy / killswitchPassesPolicy"]
```

</a>
</details>

<sub>Reviews (4): Last reviewed commit: ["test: neutral rationale comment on empty..."](https://github.com/cortexkit/anthropic-auth/commit/fc8c7b79feec6bc0a8c941afed1c75552d5c66b6) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41802610)</sub>

### CODE REVIEW

File: packages/opencode/src/sidebar-state.ts [hunk 1]
```diff
@@ -194,21 +244,68 @@ export function resolveActiveAccount(state: SidebarState): {
...
+        )
+        .join(' · ')}`
+    : ''
+
   return {
     fiveHourUsedPercent,
     sevenDayUsedPercent,
-    text: `5h: ${fiveHourUsedPercent == null ? '—' : `${Math.round(fiveHourUsedPercent)}%`} 7d: ${sevenDayUsedPercent == null ? '—' : `${Math.round(sevenDayUsedPercent)}%`}`,
+    text: visibleScoped.length ? `${text}${scopedSuffix}` : text,
   }
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/opencode/src/sidebar-state.ts]: Fixed in 44757b9. `getCollapsedQuotaSummary` now builds the `5h/7d` primary segment only when at least one of the two windows is present; when both are absent (only reachable past the early-return because a scoped window is visible), the text is just the scoped segment — no `5h: — 7d: —` prefix and no leading separator. The normal all-three case and the existing partial-window dash case (e.g. `5h: 23% 7d: —`) are byte-identical. Added a red-first test (`omits 5h/7d placeholders when only scoped windows are visible`) plus a guard test for the partial-dash case.

File: packages/core/src/accounts.ts [hunk 2]
```diff
@@ -423,6 +464,46 @@ function normalizeQuota(value: unknown): OAuthAccount['quota'] {
...
+        checkedAt,
+        resetsAt:
+          typeof entry.resetsAt === 'string' ? entry.resetsAt : undefined,
+        severity:
+          typeof entry.severity === 'string' ? entry.severity : undefined,
+        isActive:
+          typeof entry.isActive === 'boolean' ? entry.isActive : undefined,
+      })
+    }
+    if (scoped.length) quota.scoped = scoped
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/accounts.ts]: Fixed in b149c3b — good catch, this was a real contract violation on the load path. `normalizeQuota` now assigns `quota.scoped = scoped` unconditionally (inside the existing `Array.isArray(value.scoped)` guard, so pre-feature data with no `scoped` key is still not synthesized one). Added two round-trip tests that failed red before the fix: an empty `scoped: []` on a fallback account with no `five_hour`/`seven_day` (which previously dropped the entire quota to `undefined`) and one alongside a `five_hour` window — both now survive save → `loadAccounts`.

File: packages/opencode/src/tui.tsx [hunk 3]
```diff
@@ -294,6 +297,47 @@ function QuotaRow(props: {
...
+    isScopedQuotaCritical(props.window)
+      ? 'err'
+      : usageTone(used(), props.appearance)
+  return (
+    <box width='100%' flexDirection='row' justifyContent='space-between'>
+      <box flexDirection='row'>
+        <text fg={props.theme.textMuted}>
+          {`${props.window.model} only`.padEnd(8)}
+        </text>
+        <For each={quotaBarSegments(used(), props.appearance, null)}>
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/opencode/src/tui.tsx]: Fixed in b149c3b. `quotaBarSegments` now takes an optional `fillTone` override (default omitted → byte-identical to the current percentage-only behavior, so the 5h/7d `QuotaRow` path is unchanged). `ScopedQuotaRow` passes `fillTone={tone()}` so the bar fill and the percentage text use the same `isScopedQuotaCritical`-aware tone — a `severity:"critical"` window under 90% now renders a critical bar and a critical percentage together.

File: packages/opencode/src/sidebar-state.ts [hunk 4]
```diff
@@ -65,13 +75,46 @@ function isRecord(v: unknown): v is Record<string, unknown> {
...
+}
+
 export function normalizeSidebarState(raw: unknown): SidebarState {
   if (!isRecord(raw)) return { ...DEFAULT_SIDEBAR_STATE }
 
   const main: SidebarState['main'] = { quota: null }
   if (isRecord(raw.main)) {
     const m = raw.main
-    main.quota = isRecord(m.quota) ? (m.quota as AccountQuota) : null
+    main.quota = isRecord(m.quota)
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/opencode/src/sidebar-state.ts]: Fixed in b149c3b — extracted `normalizeAccountQuota(raw)` (mirrors `normalizeScopedArray`'s style) and call it from both the main and fallback branches of `normalizeSidebarState`, so the scoped projection has one source of truth.

File: packages/core/src/tests/scoped-quota.test.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+          scope: null,
+          is_active: true,
+        },
+      ],
+      1,
+    )
+    expect(out).toEqual([])
+  })
+
+  test('dedupes by trimmed id (or display_name when id is null/empty)', () => {
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/tests/scoped-quota.test.ts]: Strengthened in b149c3b — added a case with the same `id` but different `display_name`s (collapses to one, proving id-keying not name-keying) and cases with null/empty ids (fallback to the display_name slug — same name collapses, distinct names kept separately).

File: packages/core/src/tests/scoped-quota.test.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
+import { describe, expect, test } from 'bun:test'
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/tests/scoped-quota.test.ts]: Added in b149c3b — a negative case with `kind:"weekly_scoped"` but `group:"monthly"`/`"daily"` → `[]`, so the `group === "weekly"` clause is no longer a no-op in the suite.

File: packages/core/src/tests/scoped-quota.test.ts [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+import {
+  type AccountScopedQuotaWindow,
+  fetchOAuthQuotaSnapshot,
+  mapScopedWindows,
+  type OAuthLimitEntry,
+} from '../accounts.ts'
+
+const RESET = '2026-07-07T09:59:59Z'
+
+function livePayload(): { limits: OAuthLimitEntry[] } {
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/tests/scoped-quota.test.ts]: Added in b149c3b — a coexistence case with `five_hour` + `seven_day` + `limits[]` all present in one snapshot, asserting the legacy windows' values survive alongside `scoped[]`. A regression that dropped the flat windows when `limits[]` is present would now fail.
